File: src/commonTest/kotlin/com/scoreo/infrastructure/InMemoryCloudSyncRepository.kt
package com.scoreo.infrastructure

import com.scoreo.domain.port.CloudSyncRepository
import com.scoreo.domain.port.SyncData
import com.scoreo.domain.port.SyncException
import com.scoreo.domain.port.SyncStatus

class InMemoryCloudSyncRepository : CloudSyncRepository {
    var storedData: SyncData? = null
    var connected: Boolean = false
    var email: String? = null
    private var _lastSync: Long? = null
    private var _isOnline: Boolean = true
    var failNext: (() -> Exception)? = null

    fun setOnline(online: Boolean) { _isOnline = online }
    fun setLastSync(ts: Long) { _lastSync = ts }

    override suspend fun push(data: SyncData) {
        failNext?.invoke()?.let { throw it }
        if (!connected) throw SyncException.NotAuthenticated()
        storedData = data
        _lastSync = data.lastModified
    }

    override suspend fun pull(): SyncData {
        failNext?.invoke()?.let { throw it }
        if (!connected) throw SyncException.NotAuthenticated()
        return storedData ?: SyncData(
            players = emptyList(),
            gameTypes = emptyList(),
            matches = emptyList(),
            lastModified = 0L,
        )
    }

    override suspend fun getStatus(): SyncStatus = SyncStatus(
        connected = connected,
        lastSync = _lastSync,
        email = email,
        isOnline = _isOnline,
    )

    override suspend fun login() {
        connected = true
        email = "test@example.com"
    }

    override suspend fun logout() {
        connected = false
        email = null
    }
}

File: apps/scoreo/src/infrastructure/testing/inMemoryCloudSyncRepository.ts
import type {
  CloudSyncRepository,
  SyncData,
  SyncException,
  SyncStatus,
} from '../../domain/port/cloudSyncRepository'

const notAuthenticated = (): SyncException => ({ kind: 'NotAuthenticated', message: 'Not authenticated' })

export class InMemoryCloudSyncRepository implements CloudSyncRepository {
  storedData: SyncData | undefined
  connected = false
  /** Optional hook: return a SyncException to make the next push/pull throw it. */
  failNext: (() => SyncException | undefined) | undefined

  private lastSync: number | null = null
  private isOnline = true

  setOnline(online: boolean): void {
    this.isOnline = online
  }

  setLastSync(ts: number): void {
    this.lastSync = ts
  }

  async push(data: SyncData): Promise<void> {
    const failure = this.failNext?.()
    if (failure) throw failure
    if (!this.connected) throw notAuthenticated()
    this.storedData = data
    this.lastSync = data.lastModified
  }

  async pull(): Promise<SyncData> {
    const failure = this.failNext?.()
    if (failure) throw failure
    if (!this.connected) throw notAuthenticated()
    return this.storedData ?? { players: [], gameTypes: [], matches: [], lastModified: 0 }
  }

  async getStatus(): Promise<SyncStatus> {
    return {
      connected: this.connected,
      lastSync: this.lastSync,
      isOnline: this.isOnline,
    }
  }

  async login(): Promise<void> {
    this.connected = true
  }

  async logout(): Promise<void> {
    this.connected = false
  }
}
